### PR TITLE
chore(node): declutter NodeJS namespace

### PR DESCRIPTION
**Pull request by SimonSchick**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

If removing a declaration:
- [x] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [x] Delete the package's directory.
- [x] Add it to `notNeededPackages.json`.

This PR aims to declutter the `NodeJS` namespace and `globa.d.ts` by moving definitions into their respective modules.
This includes: `events`, `console`, `buffer`, `domain`.

## Events

There used to (and still is) `NodeJS.EventEmitter` this type was only created so it could be used for other global definitions, `EventEmitter` should be imported from `events` instead, a compability alias has been provided.

## Process
All process related definitions have been moved to `process.d.ts`, most popular definitions have compatilibity aliases (eg. `NodeJS.ProcessEnv` etc.), all others have been moved to `process` and must be imported from there.

~~Caveat: I can't seem to augment `HRTime` for bigint usage anymore~~ Fixed.

## Stream
As evident by my refactors to other packages, there is a general confusion between `NodeJS.WriteStream/ReadStream` (the TTY variant) and `NodeJS.ReadableStream/WritatableStream` (pure stream).
I have *removed* the TTY aliases and changed all imports in third party packages to use `tty` instead.

## Buffer
No visible changes for consumers.

There appears to be a bug in TS that causes `NodeJS.ArrayBufferView` to not be resolved correctly, I'm looking into filing an issue for this.

## Domain
By moving process definitions into their respective module, `NodeJS.Domain` is no longer needed and was *removed*, use `import { Domain } from 'domain';` instead.

# Console
`NodeJS.ConsoleConstructorOptions` and `NodeJS.ConsoleConstructor` have been moved into the `console` module.

# For the future

- Remove all DT references to `NodeJS.EventEmitter` and eventually deprecate it?
- Consolidate NodeJS.WritableStream/ReadableStream with definitions in `stream` to avoid confusion.

### CODE REVIEW

**Comment by SimonSchick** [comment]
@Flarna tagging ya for early 👀, I also need some help with the error showing up in `buffer.d.ts`.

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## derhuerst__cli-on-key ([<kbd>unpkg</kbd>](https://unpkg.com/browse/derhuerst__cli-on-key@latest/))
was missing the following properties:
1. parse
## hapi-pino ([<kbd>unpkg</kbd>](https://unpkg.com/browse/hapi-pino@latest/))
was missing the following properties:
1. name
## pkgcloud ([<kbd>unpkg</kbd>](https://unpkg.com/browse/pkgcloud@latest/))
was missing the following properties:
1. providers
1. providers
## signale ([<kbd>unpkg</kbd>](https://unpkg.com/browse/signale@latest/))
was missing the following properties:
1. scopeName
1. currentOptions
1. date
1. timestamp
1. filename

<details>
<summary>as well as these 9 other properties...</summary>
<p>packageConfiguration, configuration, scopeName, currentOptions, date, timestamp, filename, packageConfiguration, configuration</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 560fc46c439f4cb24333361d16da2cdb4ec11af1
</p>

**Comment by SimonSchick** [comment]
@danger-public I don't understand what this bot is doing at all.

**Comment by peterblazejewicz** [comment]
@SimonSchick
You'll probably have to name the reviews explicitely, by adding their GitHub handles. For some reason the bot failed here

for the 'dt-danger':
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44430

**Comment by SimonSchick** [comment]
Hey @Flarna

**Comment by Flarna** [comment]
@SimonSchick sorry, I'm on vacation... will take a while.

File: types/node/globals.d.ts [hunk 1]
```diff
@@ -549,439 +150,11 @@ declare namespace NodeJS {
...
-        listeners(event: "newListener"): NewListenerListener[];
-        listeners(event: "removeListener"): RemoveListenerListener[];
-        listeners(event: "multipleResolves"): MultipleResolveListener[];
-    }
-
     interface Global {
         Array: typeof Array;
         ArrayBuffer: typeof ArrayBuffer;
+        Buffer: typeof Buffer; // TODO: Move to `buffer.d.ts` at ts3.9
         Boolean: typeof Boolean;
```
Comment (ExE-Boss) [inline types/node/globals.d.ts:157]: This should remain in alphabetical order:

```suggestion
        Boolean: typeof Boolean;
        Buffer: typeof Buffer; // TODO: Move to `buffer.d.ts` at ts3.9
```

File: types/node/events.d.ts [hunk 2]
```diff
@@ -35,9 +35,6 @@ declare module "events" {
          */
         let captureRejections: boolean;
 
-        interface EventEmitter extends NodeJS.EventEmitter {
-        }
-
         class EventEmitter {
```
Comment (ExE-Boss) [inline types/node/events.d.ts:38]: For backwards compatibility, `'events'.EventEmitter` should keep extending `NodeJS.EventEmitter`:
```suggestion
        interface EventEmitter extends NodeJS.EventEmitter {
        }

        class EventEmitter {
```

File: types/node/events.d.ts [hunk 3]
```diff
@@ -53,6 +50,30 @@ declare module "events" {
...
+            prependListener(event: string | symbol, listener: (...args: any[]) => void): this;
+            prependOnceListener(event: string | symbol, listener: (...args: any[]) => void): this;
+            eventNames(): Array<string | symbol>;
+        }
+    }
+
+    global {
+        namespace NodeJS {
+            interface EventEmitter extends EventEmitter.EventEmitter {
+            }
```
Comment (ExE-Boss) [inline types/node/events.d.ts:76]:
```suggestion
        }
    }

    global {
        namespace NodeJS {
            interface EventEmitter {
                addListener(event: string | symbol, listener: (...args: any[]) => void): this;
                on(event: string | symbol, listener: (...args: any[]) => void): this;
                once(event: string | symbol, listener: (...args: any[]) => void): this;
                removeListener(event: string | symbol, listener: (...args: any[]) => void): this;
                off(event: string | symbol, listener: (...args: any[]) => void): this;
                removeAllListeners(event?: string | symbol): this;
                setMaxListeners(n: number): this;
                getMaxListeners(): number;
                listeners(event: string | symbol): Function[];
                rawListeners(event: string | symbol): Function[];
                emit(event: string | symbol, ...args: any[]): boolean;
                listenerCount(type: string | symbol): number;
                // Added in Node 6...
                prependListener(event: string | symbol, listener: (...args: any[]) => void): this;
                prependOnceListener(event: string | symbol, listener: (...args: any[]) => void): this;
                eventNames(): Array<string | symbol>;
            }
```

File: types/node/events.d.ts (hunk 2 above)
Comment (SimonSchick) [reply types/node/events.d.ts:38]: This should not matter as they still result in the same effective type definition.
I haven't found *any* packages that had issues with the reversal in inheritance (hence no NodeJS.EventEmitter related changes in other packages).

Comment (SimonSchick) [reply types/node/events.d.ts:38]: That said if you can point to a potential issue I'm willing to change my mind.

**Review by orta — APPROVED** [review]
This looks good to me, I've just asked in our DT chat room for one more look over it

**Comment by orta** [comment]
OK, looks like we're a bit skeptical of a change this big not breaking quite a few builds (the Buffer changes, and interface merging being prominent examples) with the recommendations being either:

 - break it into smaller PRs which we can merge over time
 - ship this change with the next major node versions

**Comment by SimonSchick** [comment]
Can you elaborate on the buffer changes here?

**Comment by SimonSchick** [comment]
@orta also which chat room, is it https://gitter.im/DefinitelyTyped/DefinitelyTyped?

**Comment by SimonSchick** [comment]
@orta ping

In any case I'm fine with holding off these rather sweeping changes till v15 hits.

**Comment by orta** [comment]
Sorry, my GH notifications for the DT org are basically useless when I'm on rotation. 

The idea was that we'd expect there to be a bunch of broken apps and libraries breaking due to these changes, and while the idea is good making it a part of a major bump means people will be OK with that

**Comment by ExE-Boss** [comment]
@orta Looks like the bot got stuck in an infinite loop, again.

**Comment by SimonSchick** [comment]
Closing for now, will reopen with v15.

**Pull request closed without merging**